File: ruby/beer-song/beer_song.rb
class BeerSong
  def verse(n)
    b = Beer.bottles(n)
    result = ''
    result << b.measure.capitalize << ' of beer on the wall, '
    result << b.measure << ' of beer.' << "\n"
    result << b.action << ', '
    result << b.pred.measure << ' of beer on the wall.' << "\n"
    result
  end

  def verses(start, finish)
    start.downto(finish).map { |n| verse(n) }.join("\n") + "\n"
  end

  def sing
    verses(99, 0)
  end
end

module Beer
  def self.bottles(n)
    { 0 => Zero.new,
      1 => One.new
    }.fetch(n) { Number.new(n) }
  end

  class Number
    attr_reader :n

    def initialize(n)
      @n = n
    end

    def pronoun
      'one'
    end

    def action
      "Take #{pronoun} down and pass it around"
    end

    def measure
      "#{n} bottles"
    end

    def pred
      Beer.bottles(n - 1)
    end
  end

  class Zero < Number
    def initialize
      super(0)
    end

    def action
      'Go to the store and buy some more'
    end

    def measure
      'no more bottles'
    end

    def pred
      Beer.bottles(99)
    end
  end

  class One < Number
    def initialize
      super(1)
    end

    def pronoun
      'it'
    end

    def measure
      "#{n} bottle"
    end
  end
end

File: ecmascript/beer-song/beer-song.js
const bottles = (n) => {
  switch (n) {
    case 0: return `no more bottles`;
    case -1: return '99 bottles';
    default: return `${n} bottle${n==1 ? '' : 's'}`
  }
};

const capitalize = (s) => {
  return s.charAt(0).toUpperCase() + s.slice(1);
};

const takeDown = (n) => {
  switch (n) {
    case 0: return 'Go to the store and buy some more';
    case 1: return 'Take it down and pass it around';
    default: return 'Take one down and pass it around';
  }
};

const start = (n) => {
  return `${capitalize(bottles(n))} of beer on the wall, ${bottles(n)} of beer.`;
};

const end = (n) => {
  return `${takeDown(n)}, ${bottles(n-1)} of beer on the wall.`;
};

class Beer {
  static verse(n) {
    return [].concat(start(n)).concat(end(n)).concat('').join('\n');
  }

  static sing(a = 99, b = 0) {
    const verses = [];
    for (let i=a; i>=b; i--) {
      verses.push(this.verse(i));
    }
    return verses.join('\n');
  }
}

export default Beer;
